### PR TITLE
[core] fix unexpected chain id (#4129)

### PR DESCRIPTION
**Pull request by GheisMohammadi**
fix unexpected chain id by adding a fork logic for EIP-1344 to return the Ethereum compatible chain id which is retrieved from chain config

## Issue
https://github.com/harmony-one/harmony/issues/4129

### CODE REVIEW

File: internal/params/config.go [hunk 1]
```diff
@@ -98,6 +99,7 @@ var (
 		SHA3Epoch:                  big.NewInt(74570),
 		HIP6And8Epoch:              big.NewInt(74570),
 		StakingPrecompileEpoch:     big.NewInt(75175),
+		ChainIdFixEpoch:            big.NewInt(75175),
```
Comment (rlan35) [inline internal/params/config.go:102]: Set it to be future epoch. Forking can be not retrospective.

File: internal/params/config.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ var (
 		SHA3Epoch:                  big.NewInt(725), // Around Mon Oct 11 2021, 19:00 UTC
 		HIP6And8Epoch:              big.NewInt(725), // Around Mon Oct 11 2021, 19:00 UTC
 		StakingPrecompileEpoch:     big.NewInt(871), // Around Tue Feb 11 2022
+		ChainIdFixEpoch:            big.NewInt(871), // Around Tue Feb 11 2022
```
Comment (rlan35) [inline internal/params/config.go:68]: Set it to be TDBEpoch. We haven't decided when to launch on mainnet yet.

File: internal/params/config.go (hunk 1 above)
Comment (GheisMohammadi) [reply internal/params/config.go]: I changed the epoch to 75685.

File: internal/params/config.go [hunk 3]
```diff
@@ -98,6 +99,7 @@ var (
 		SHA3Epoch:                  big.NewInt(74570),
 		HIP6And8Epoch:              big.NewInt(74570),
 		StakingPrecompileEpoch:     big.NewInt(75175),
+		ChainIdFixEpoch:            big.NewInt(75685),
```
Comment (MaxMustermann2) [inline internal/params/config.go:102]: Please change this to `EpochTBD` too, and merge with branch `main`. Thank you.

Comment (GheisMohammadi) [reply internal/params/config.go]: it's done, thanks

**Comment by GheisMohammadi** [comment]
> 
Basically solidity developers use chainID to get the Ethereum compatible chain id. The fixes we made, returns ethereum compatible chain id for EIP-1344 which affects the EVM and smart contracts. So, it needs to be hard fork logic. The another option is to add one more opcode to EVM (like ethChainID ) to return ethereum compatible chain id. So, the solidity developers have to know either use chainID or ethChainID to retrieve the proper chain id. But this solution will break the Ethereum compatibility.  
It won't affect RPC call, because RPC retrieves chainID from harmony object which is reading directly from chainDB.

**Review by rlan35 — APPROVED** [review]

File: internal/params/config.go [hunk 4]
```diff
@@ -268,6 +274,7 @@ var (
 		big.NewInt(0),                      // SHA3Epoch
 		big.NewInt(0),                      // HIP6And8Epoch
 		big.NewInt(0),                      // StakingPrecompileEpoch
+		big.NewInt(0),						// ChainIdFixEpoch
```
Comment (rlan35) [inline internal/params/config.go:277]: please align the comment with others.

Comment (GheisMohammadi) [reply internal/params/config.go]: done

**Review by MaxMustermann2 — APPROVED** [review]

**Comment by MaxMustermann2** [comment]
Please resolve conflicts with `internal/params/config.go`

**Comment by MaxMustermann2** [comment]
Please run `goimports -w -e internal/params/config.go` and commit to fix the build.

**Review by MaxMustermann2 — APPROVED** [review]

**Comment by sophoah** [comment]
what kind of test can we run here to confirm it works ?

**Comment by GheisMohammadiBC** [comment]
It can be tested by a simple smart contract with a view function to retrieve chainID.

**Comment by GheisMohammadi** [comment]
> what kind of test can we run here to confirm it works ?

It can be tested by a simple smart contract with a view function to retrieve chainID.

**Comment by MaxMustermann2** [comment]
Here's the Solidity contract
```solidity
//SPDX-License-Identifier: Unlicense
pragma solidity ^0.8.0;

contract SampleContract {
    function sampleFunction() external view returns (uint256) {
        uint256 id;
        assembly {
            id := chainid()
        }
        return id;
    }
}
```

Here's the localnet steps to run it
```python
$ brownie console --network hmyLocal
>>> accounts.add( '1f84c95ac16e6a50f08d44c7bde7aff8742212fda6e4321fde48bf83bef266dc' )
<LocalAccount '0xA5241513DA9F4463F1d4874b548dFBAC29D91f34'>
>>> SampleContract.deploy( { 'from': accounts[ 0 ] } )
Transaction sent: 0x2c2df26e9980d33cb143ef5c1394787f9e217d41d00c13eb34bf46da5a0288e9
  Gas price: 30.0 gwei   Gas limit: 86824   Nonce: 1
  SampleContract.constructor confirmed   Block: 88   Gas used: 78931 (90.91%)
  SampleContract deployed at: 0x8BcF541eA0d90a4001F5B3f6667c495aBcBFC9A2
>>> contract = SampleContract[ -1 ]
>>> contract.sampleFunction()
1666700000
>>> web3.eth.chainId
1666700000
```

**Review by rlan35 — APPROVED** [review]

**Comment by rlan35** [comment]
please resolve conflict so we can merge

**Review by rlan35 — APPROVED** [review]

**Pull request merged**